src/HatsSignerGate.sol

```diff
@@ -407,14 +413,19 @@ contract HatsSignerGate is
     // ensure that the call is coming from the safe
     if (msg.sender != address(safe)) revert NotCalledFromSafe();
 
-     /// @dev This is a reentrancy guard designed to work with the `checkAfterExecution()` function. It allows reentrancy
-    /// into this contract so that the `checkAfterExecution()` function can be called by the `safe`, but it only allows
-    /// one call each of `checkTransaction()` and `checkAfterExecution()`.
-    unchecked {
-      ++_guardEntries;
+    // prevent calling this function from execTransactionFromModule or execTransactionFromModuleReturnData
+    if (_reentrancyGuard == 1) revert NoReentryAllowed();
+
+    // record the initial nonce of the safe at the beginning of the transaction
+    if (_entrancyCounter == 0) {
+      _initialNonce = safe.nonce() - 1;
     }
-    // revert if re-entry into this function is detected prior to `checkAfterExecution()` is called
-    if (_guardEntries > 1) revert NoReentryAllowed();
+
+    // increment the entrancy count
+    _entrancyCounter++;
+
+    // prevent entrancy to this function other than by the safe
```

Comment:
elaborate on this logic more in the comments. I think its ok to err on the side of too long here given the nuance and sensitivity.

Whenever this function is called from a source other than Safe.execTransaction, its possible that it is a malicious call attempting to manipulate the transient storage so that an attacker can make malicious changes to the Safe state without detection by this guard in IGuard.checkAfterExecution. To prevent this, we rely on the invariant that the Safe nonce increments every time Safe.execTransaction calls out to IGuard.checkTransaction, allowing us to ensure that this function is only called the same number of times in a single transaction as Safe.execTransaction calls, ie by how much the Safe's nonce increases.